File: tools/jitterbug/src/main/java/org/pushingpixels/tools/jitterbug/JColorSchemeList.java
/*
 * Copyright (c) 2005-2018 Substance Kirill Grouchnikov. All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  o Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *
 *  o Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 *  o Neither the name of Substance Kirill Grouchnikov nor the names of
 *    its contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.pushingpixels.tools.jitterbug;

import org.pushingpixels.substance.api.SubstanceSkin;
import org.pushingpixels.substance.api.colorscheme.*;
import org.pushingpixels.substance.api.renderer.SubstanceDefaultListCellRenderer;
import org.pushingpixels.tools.jitterbug.svg.*;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import javax.swing.event.ListSelectionEvent;
import javax.swing.filechooser.FileNameExtensionFilter;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.io.*;
import java.util.regex.*;

public class JColorSchemeList extends JComponent {
    static final String LIST = "List";
    static final String INSTRUCTIONAL = "Instructional";
    private SubstanceSkin.ColorSchemes schemes;
    private JList<SubstanceColorScheme> schemeList;
    private JPanel cardPanel;
    private ColorSchemeListModel schemeListModel;
    private File schemesFile;

    private boolean isModified;

    private JButton moveUpButton;

    private JButton moveDownButton;

    private JButton renameButton;

    private JButton deleteButton;

    public JColorSchemeList() {
        this.schemes = new SubstanceSkin.ColorSchemes();

        this.schemeList = new JList<>();
        this.schemeListModel = new ColorSchemeListModel();
        this.schemeList.setModel(schemeListModel);
        this.schemeList.setCellRenderer(new ColorSchemeListRenderer());
        this.schemeList.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
        this.schemeList.getSelectionModel().addListSelectionListener((ListSelectionEvent e) -> {
            Object selected = schemeList.getSelectedValue();
            firePropertyChange("selectedColorScheme", null, selected);
        });

        JPanel bottomButtonsPanel = new JPanel(new FlowLayout(FlowLayout.TRAILING));
        JButton addColorScheme = new JButton();
        addColorScheme.setToolTipText("Adds a new color scheme");
        addColorScheme.setIcon(outline_add_24px.of(12, 12));
        addColorScheme.addActionListener((ActionEvent e) -> SwingUtilities.invokeLater(() -> {
            String newName = getNewColorSchemeName(null);
            if (newName == null) {
                return;
            }

            SubstanceColorScheme newScheme = new BaseLightColorScheme(newName) {
                @Override
                public Color getForegroundColor() { return Color.white; }

                @Override
                public Color getUltraLightColor() { return Color.white; }

                @Override
                public Color getExtraLightColor() { return Color.white; }

                @Override
                public Color getLightColor() { return Color.white; }

                @Override
                public Color getMidColor() { return Color.white; }

                @Override
                public Color getDarkColor() { return Color.white; }

                @Override
                public Color getUltraDarkColor() { return Color.white; }
            };

            schemes.add(newScheme);
            schemeListModel.fireContentsChanged();
            schemeList.setSelectedValue(newScheme, true);
            setModified(true);
        }));

        bottomButtonsPanel.add(addColorScheme);

        bottomButtonsPanel.add(Box.createHorizontalStrut(10));

        moveUpButton = new JButton();
        moveUpButton.setToolTipText("Moves the currently selected color scheme up");
        moveUpButton.setEnabled(false);
        moveUpButton.setIcon(outline_keyboard_arrow_up_24px.of(12, 12));
        moveUpButton.addActionListener((ActionEvent e) -> SwingUtilities.invokeLater(() -> {
            SubstanceColorScheme selected = (SubstanceColorScheme) schemeList.getSelectedValue();
            schemes.switchWithPrevious(selected.getDisplayName());
            schemeListModel.fireContentsChanged();
            schemeList.setSelectedValue(selected, true);
            setModified(true);
        }));
        bottomButtonsPanel.add(moveUpButton);

        moveDownButton = new JButton();
        moveDownButton.setToolTipText("Moves the currently selected color scheme down");
        moveDownButton.setEnabled(false);
        moveDownButton.setIcon(outline_keyboard_arrow_down_24px.of(12, 12));
        moveDownButton.addActionListener((ActionEvent e) -> SwingUtilities.invokeLater(() -> {
            SubstanceColorScheme selected = (SubstanceColorScheme) schemeList.getSelectedValue();
            schemes.switchWithNext(selected.getDisplayName());
            schemeListModel.fireContentsChanged();
            schemeList.setSelectedValue(selected, true);
            setModified(true);
        }));
        bottomButtonsPanel.add(moveDownButton);

        bottomButtonsPanel.add(Box.createHorizontalStrut(10));

        renameButton = new JButton();
        renameButton.setToolTipText("Renames the currently selected color scheme");
        renameButton.setEnabled(false);
        renameButton.setIcon(outline_edit_24px.of(12, 12));
        renameButton.addActionListener((ActionEvent e) -> SwingUtilities.invokeLater(() -> {
            SubstanceColorScheme selected = (SubstanceColorScheme) schemeList.getSelectedValue();
            String newName = getNewColorSchemeName(selected.getDisplayName());
            if (newName == null)
                return;

            if (newName.equals(selected.getDisplayName()))
                return;

            boolean isLight = !selected.isDark();

            SubstanceColorScheme renamedScheme = isLight
                    ? new BaseLightColorScheme(newName) {
                        @Override
                        public Color getForegroundColor() { return selected.getForegroundColor(); }

                        @Override
                        public Color getUltraLightColor() { return selected.getUltraLightColor(); }

                        @Override
                        public Color getExtraLightColor() { return selected.getExtraLightColor(); }

                        @Override
                        public Color getLightColor() { return selected.getLightColor(); }

                        @Override
                        public Color getMidColor() { return selected.getMidColor(); }

                        @Override
                        public Color getDarkColor() { return selected.getDarkColor(); }

                        @Override
                        public Color getUltraDarkColor() { return selected.getUltraDarkColor(); }
                    } : new BaseDarkColorScheme(newName) {
                        @Override
                        public Color getForegroundColor() { return selected.getForegroundColor(); }

                        @Override
                        public Color getUltraLightColor() { return selected.getUltraLightColor(); }

                        @Override
                        public Color getExtraLightColor() { return selected.getExtraLightColor(); }

                        @Override
                        public Color getLightColor() { return selected.getLightColor(); }

                        @Override
                        public Color getMidColor() { return selected.getMidColor(); }

                        @Override
                        public Color getDarkColor() { return selected.getDarkColor(); }

                        @Override
                        public Color getUltraDarkColor() { return selected.getUltraDarkColor(); }
                    };

            schemes.replace(selected.getDisplayName(), renamedScheme);
            schemeListModel.fireContentsChanged();
            schemeList.getSelectionModel().clearSelection();
            schemeList.setSelectedValue(renamedScheme, true);
            setModified(true);
        }));

        bottomButtonsPanel.add(renameButton);

        deleteButton = new JButton();
        deleteButton.setToolTipText("Removes the currently selected color scheme");
        deleteButton.setEnabled(false);
        deleteButton.setIcon(outline_delete_24px.of(12, 12));
        deleteButton.addActionListener((ActionEvent e) -> SwingUtilities.invokeLater(() -> {
            SubstanceColorScheme selected = (SubstanceColorScheme) schemeList.getSelectedValue();
            schemes.delete(selected.getDisplayName());
            schemeListModel.fireContentsChanged();
            schemeList.getSelectionModel().clearSelection();
            setModified(true);
        }));

        bottomButtonsPanel.add(deleteButton);

        this.schemeList.getSelectionModel().addListSelectionListener((ListSelectionEvent e) -> {
            int selectedIndex = schemeList.getSelectedIndex();
            boolean hasSelection = (selectedIndex >= 0);
            renameButton.setEnabled(hasSelection);
            deleteButton.setEnabled(hasSelection);

            moveUpButton.setEnabled(hasSelection && (selectedIndex > 0));
            moveDownButton.setEnabled(hasSelection && (selectedIndex < (schemes.size() - 1)));
        });

        JPanel mainControlPanel = new JPanel(new BorderLayout());
        mainControlPanel.add(new JScrollPane(this.schemeList), BorderLayout.CENTER);
        mainControlPanel.add(bottomButtonsPanel, BorderLayout.SOUTH);

        // Set up a card panel. It has two cards:
        // * instructional text
        // * list of color schemes with button controls
        this.cardPanel = new JPanel(new CardLayout());
        JTextArea textArea = new JTextArea(
                "List of color schemes. Use one of the following:\n   * Drag a .colorschemes file" +
                        " here\n   * Click the \"New\" button below");
        textArea.setBorder(new EmptyBorder(10, 8, 0, 8));
        textArea.setEditable(false);
        textArea.setOpaque(false);
        textArea.setRows(9);
        // textArea.setColumns(20);
        this.cardPanel.add(textArea, INSTRUCTIONAL);
        this.cardPanel.add(mainControlPanel, LIST);

        this.setLayout(new BorderLayout());
        this.add(this.cardPanel, BorderLayout.CENTER);

        // constrain the preferred size so that the layout doesn't
        // change once the control panel is shown
        this.setPreferredSize(textArea.getPreferredSize());
    }

    public void setColorSchemeList(File file) {
        if (!this.checkModifiedStateAndSaveIfNecessary()) {
            return;
        }
        try {
            this.schemesFile = file;
            if (file != null) {
                this.schemes = SubstanceSkin.getColorSchemes(file.toURI().toURL());
            } else {
                this.schemes = new SubstanceSkin.ColorSchemes();
            }
            // switch to the list view
            CardLayout cl = (CardLayout) (cardPanel.getLayout());
            cl.show(cardPanel, LIST);
            // let the view know that the model has changed
            schemeListModel.fireContentsChanged();
            // update the buttons state
            if (file == null) {
                moveUpButton.setEnabled(false);
                moveDownButton.setEnabled(false);
                deleteButton.setEnabled(false);
                renameButton.setEnabled(false);
            }
            this.setModified(false);
        } catch (Exception exc) {
            exc.printStackTrace();
        }
    }

    public boolean checkModifiedStateAndSaveIfNecessary() {
        if (!this.isModified) {
            return true;
        }

        // does the user want to save modifications?
        String fileName = (this.schemesFile == null) ? "Unsaved" : this.schemesFile.getName();
        int userSelection = JOptionPane.showConfirmDialog(SwingUtilities.getWindowAncestor(this),
                "Do you want to save the changes to '" + fileName + "'?", "Modified contents",
                JOptionPane.YES_NO_CANCEL_OPTION);
        if (userSelection == JOptionPane.CANCEL_OPTION) {
            return false;
        }
        if (userSelection == JOptionPane.YES_OPTION) {
            if (this.schemesFile != null) {
                // ask to save
                this.save();
            } else {
                this.saveAs();
            }
        }
        return true;
    }

    private String getFileNameForSaving() {
        JFileChooser jfc = new JFileChooser();
        FileNameExtensionFilter filter = new FileNameExtensionFilter("Color scheme files",
                "colorschemes");
        jfc.setFileFilter(filter);
        int returnVal = jfc.showSaveDialog(SwingUtilities.getWindowAncestor(this));
        if (returnVal == JFileChooser.APPROVE_OPTION) {
            String res = jfc.getSelectedFile().getAbsolutePath();
            if (!res.endsWith(".colorschemes"))
                res += ".colorschemes";
            return res;
        }
        return null;
    }

    private String getNewColorSchemeName(String nameToStartWith) {
        String result = JOptionPane.showInputDialog(SwingUtilities.getWindowAncestor(this),
                "Type the color scheme name", nameToStartWith);
        if (result == null)
            return null;

        if (result.equals(nameToStartWith))
            return result;

        result = result.trim();

        // check characters
        Pattern pattern = Pattern.compile("[a-zA-Z ]+");
        Matcher matcher = pattern.matcher(result);
        if (!matcher.matches()) {
            JOptionPane.showMessageDialog(SwingUtilities.getWindowAncestor(this),
                    "Only use letters and spaces", "Name invalid", JOptionPane.ERROR_MESSAGE);
            return null;
        }

        // check for uniqueness
        SubstanceColorScheme existing = this.schemes.get(result);
        if (existing != null) {
            JOptionPane.showMessageDialog(SwingUtilities.getWindowAncestor(this),
                    "Name already exists", "Name clash", JOptionPane.ERROR_MESSAGE);
            return null;
        }
        return result;
    }

    public void save() {
        this.saveAs(this.schemesFile);
    }

    public void saveAs() {
        // ask for a filename for the save
        String fileName = this.getFileNameForSaving();
        if (fileName != null) {
            File file = new File(fileName);
            this.saveAs(file);
        }
    }

    private void saveAs(File file) {
        PrintStream printStream = null;

        try {
            printStream = new PrintStream(new FileOutputStream(file));
            for (int i = 0; i < this.schemes.size(); i++) {
                SubstanceColorScheme colorScheme = this.schemes.get(i);
                String encodedColorScheme = colorScheme.toString();
                printStream.println(encodedColorScheme);
                printStream.println();
            }
        } catch (IOException ioe) {
            ioe.printStackTrace();
        }
        if (printStream != null) {
            printStream.close();
        }
        this.schemesFile = file;
        this.setModified(false);
    }

    public File getCurrentFile() {
        return this.schemesFile;
    }

    public void updateColorScheme(SubstanceColorScheme colorScheme) {
        SubstanceColorScheme existing = this.schemes.get(colorScheme.getDisplayName());
        if (existing == null)
            throw new IllegalArgumentException(
                    "Color scheme " + colorScheme.getDisplayName() + " not found");

        this.schemes.replace(colorScheme.getDisplayName(), colorScheme);
    }

    public boolean isModified() {
        return isModified;
    }

    public void setModified(boolean isModified) {
        if (this.isModified == isModified)
            return;

        boolean old = this.isModified;
        this.isModified = isModified;
        this.firePropertyChange("modified", old, isModified);
    }

    class ColorSchemeListModel extends AbstractListModel<SubstanceColorScheme> {
        @Override
        public int getSize() {
            return (schemes == null) ? 0 : schemes.size();
        }

        @Override
        public SubstanceColorScheme getElementAt(int index) {
            if (schemes == null)
                return null;
            return schemes.get(index);
        }

        void fireContentsChanged() {
            this.fireContentsChanged(schemeList, 0, getSize());
        }
    }

    class ColorSchemeListRenderer extends SubstanceDefaultListCellRenderer {
        @Override
        public Component getListCellRendererComponent(JList list, Object value, int index,
                boolean isSelected, boolean cellHasFocus) {
            SubstanceColorScheme scheme = (SubstanceColorScheme) value;
            return super.getListCellRendererComponent(list, scheme.getDisplayName(), index,
                    isSelected, cellHasFocus);
        }
    }
}

File: tools/apollo/src/main/kotlin/org/pushingpixels/tools/apollo/JColorSchemeList.kt
/*
 * Copyright (c) 2005-2019 Radiance Apollo Kirill Grouchnikov. All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  o Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *
 *  o Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 *  o Neither the name of Radiance Apollo Kirill Grouchnikov nor the names of
 *    its contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.pushingpixels.tools.apollo

import org.pushingpixels.meteor.addDelayedActionListener
import org.pushingpixels.meteor.swing.*
import org.pushingpixels.substance.api.SubstanceSkin
import org.pushingpixels.substance.api.colorscheme.BaseDarkColorScheme
import org.pushingpixels.substance.api.colorscheme.BaseLightColorScheme
import org.pushingpixels.substance.api.colorscheme.SubstanceColorScheme
import org.pushingpixels.substance.api.renderer.SubstanceDefaultListCellRenderer
import org.pushingpixels.tools.apollo.svg.*
import java.awt.*
import java.io.File
import java.io.FileOutputStream
import java.io.PrintStream
import java.util.regex.Pattern
import javax.swing.*
import javax.swing.border.EmptyBorder
import javax.swing.filechooser.FileNameExtensionFilter
import kotlin.properties.Delegates

class JColorSchemeList : JComponent() {
    private var schemes: SubstanceSkin.ColorSchemes? = null
    private val schemeList: JList<SubstanceColorScheme>
    private val cardPanel: JPanel
    private val schemeListModel: ColorSchemeListModel
    var currentFile: File? = null
        private set

    var isModified: Boolean by Delegates.observable(false) {
        prop, old, new -> this.firePropertyChange(prop.name, old, new)
    }
    var selectedColorScheme: SubstanceColorScheme? by Delegates.observable<SubstanceColorScheme?>(null) {
        prop, old, new -> this.firePropertyChange(prop.name, old, new)
    }

    private val moveUpButton: JButton

    private val moveDownButton: JButton

    private val renameButton: JButton

    private val deleteButton: JButton

    private val fileNameForSaving: String?
        get() {
            val jfc = JFileChooser()
            val filter = FileNameExtensionFilter("Color scheme files", "colorschemes")
            jfc.fileFilter = filter
            val returnVal = jfc.showSaveDialog(SwingUtilities.getWindowAncestor(this))
            if (returnVal == JFileChooser.APPROVE_OPTION) {
                var res = jfc.selectedFile.absolutePath
                if (!res.endsWith(".colorschemes")) {
                    res += ".colorschemes"
                }
                return res
            }
            return null
        }

    init {
        this.schemes = SubstanceSkin.ColorSchemes()

        this.schemeList = JList()
        this.schemeListModel = ColorSchemeListModel()
        this.schemeList.model = schemeListModel
        this.schemeList.cellRenderer = ColorSchemeListRenderer()
        this.schemeList.selectionMode = ListSelectionModel.SINGLE_SELECTION
        this.schemeList.selectionModel.addListSelectionListener {
            selectedColorScheme = schemeList.selectedValue
        }

        val bottomButtonsPanel = JPanel(FlowLayout(FlowLayout.TRAILING))
        val addColorScheme = JButton()
        addColorScheme.toolTipText = "Adds a new color scheme"
        addColorScheme.icon = outline_add_24px.of(12, 12)
        addColorScheme.addDelayedActionListener {
            val newName = getNewColorSchemeName(null)
            if (newName != null) {
                val newScheme = object : BaseLightColorScheme(newName) {
                    override fun getForegroundColor(): Color {
                        return Color.white
                    }

                    override fun getUltraLightColor(): Color {
                        return Color.white
                    }

                    override fun getExtraLightColor(): Color {
                        return Color.white
                    }

                    override fun getLightColor(): Color {
                        return Color.white
                    }

                    override fun getMidColor(): Color {
                        return Color.white
                    }

                    override fun getDarkColor(): Color {
                        return Color.white
                    }

                    override fun getUltraDarkColor(): Color {
                        return Color.white
                    }
                }

                schemes!!.add(newScheme)
                schemeListModel.fireContentsChanged()
                schemeList.setSelectedValue(newScheme, true)
                isModified = true
            }
        }

        bottomButtonsPanel.add(addColorScheme)

        bottomButtonsPanel.add(Box.createHorizontalStrut(10))

        moveUpButton = JButton()
        moveUpButton.toolTipText = "Moves the currently selected color scheme up"
        moveUpButton.isEnabled = false
        moveUpButton.icon = outline_keyboard_arrow_up_24px.of(12, 12)
        moveUpButton.addDelayedActionListener {
            val selected = schemeList.selectedValue as SubstanceColorScheme
            schemes!!.switchWithPrevious(selected.displayName)
            schemeListModel.fireContentsChanged()
            schemeList.setSelectedValue(selected, true)
            isModified = true
        }
        bottomButtonsPanel.add(moveUpButton)

        moveDownButton = JButton()
        moveDownButton.toolTipText = "Moves the currently selected color scheme down"
        moveDownButton.isEnabled = false
        moveDownButton.icon = outline_keyboard_arrow_down_24px.of(12, 12)
        moveDownButton.addDelayedActionListener {
            val selected = schemeList.selectedValue
            schemes!!.switchWithNext(selected.displayName)
            schemeListModel.fireContentsChanged()
            schemeList.setSelectedValue(selected, true)
            isModified = true
        }
        bottomButtonsPanel.add(moveDownButton)

        bottomButtonsPanel.add(Box.createHorizontalStrut(10))

        renameButton = JButton()
        renameButton.toolTipText = "Renames the currently selected color scheme"
        renameButton.isEnabled = false
        renameButton.icon = outline_edit_24px.of(12, 12)
        renameButton.addDelayedActionListener {
            val selected = schemeList.selectedValue
            val newName = getNewColorSchemeName(selected.displayName)
            if ((newName != null) && (newName.compareTo(selected.displayName) != 0)) {

                val isLight = !selected.isDark

                val renamedScheme = if (isLight)
                    object : BaseLightColorScheme(newName) {
                        override fun getForegroundColor(): Color {
                            return selected.foregroundColor
                        }

                        override fun getUltraLightColor(): Color {
                            return selected.ultraLightColor
                        }

                        override fun getExtraLightColor(): Color {
                            return selected.extraLightColor
                        }

                        override fun getLightColor(): Color {
                            return selected.lightColor
                        }

                        override fun getMidColor(): Color {
                            return selected.midColor
                        }

                        override fun getDarkColor(): Color {
                            return selected.darkColor
                        }

                        override fun getUltraDarkColor(): Color {
                            return selected.ultraDarkColor
                        }
                    }
                else
                    object : BaseDarkColorScheme(newName) {
                        override fun getForegroundColor(): Color {
                            return selected.foregroundColor
                        }

                        override fun getUltraLightColor(): Color {
                            return selected.ultraLightColor
                        }

                        override fun getExtraLightColor(): Color {
                            return selected.extraLightColor
                        }

                        override fun getLightColor(): Color {
                            return selected.lightColor
                        }

                        override fun getMidColor(): Color {
                            return selected.midColor
                        }

                        override fun getDarkColor(): Color {
                            return selected.darkColor
                        }

                        override fun getUltraDarkColor(): Color {
                            return selected.ultraDarkColor
                        }
                    }

                schemes!!.replace(selected.displayName, renamedScheme)
                schemeListModel.fireContentsChanged()
                schemeList.selectionModel.clearSelection()
                schemeList.setSelectedValue(renamedScheme, true)
                isModified = true
            }
        }

        bottomButtonsPanel.add(renameButton)

        deleteButton = JButton()
        deleteButton.toolTipText = "Removes the currently selected color scheme"
        deleteButton.isEnabled = false
        deleteButton.icon = outline_delete_24px.of(12, 12)
        deleteButton.addDelayedActionListener {
            val selected = schemeList.selectedValue
            schemes!!.delete(selected.displayName)
            schemeListModel.fireContentsChanged()
            schemeList.selectionModel.clearSelection()
            isModified = true
        }

        bottomButtonsPanel.add(deleteButton)

        this.schemeList.selectionModel.addListSelectionListener {
            val selectedIndex = schemeList.selectedIndex
            val hasSelection = selectedIndex >= 0
            renameButton.isEnabled = hasSelection
            deleteButton.isEnabled = hasSelection

            moveUpButton.isEnabled = hasSelection && selectedIndex > 0
            moveDownButton.isEnabled = hasSelection && selectedIndex < schemes!!.size() - 1
        }

        val mainControlPanel = JPanel(BorderLayout())
        mainControlPanel.add(JScrollPane(this.schemeList), BorderLayout.CENTER)
        mainControlPanel.add(bottomButtonsPanel, BorderLayout.SOUTH)

        // Set up a card panel. It has two cards:
        // * instructional text
        // * list of color schemes with button controls
        this.cardPanel = JPanel(CardLayout())
        val textArea = JTextArea(
                "List of color schemes. Use one of the following:\n   * Drag a .colorschemes file" + " here\n   * Click the \"New\" button below")
        textArea.border = EmptyBorder(10, 8, 0, 8)
        textArea.isEditable = false
        textArea.isOpaque = false
        textArea.rows = 9
        // textArea.setColumns(20);
        this.cardPanel.add(textArea, INSTRUCTIONAL)
        this.cardPanel.add(mainControlPanel, LIST)

        this.layout = BorderLayout()
        this.add(this.cardPanel, BorderLayout.CENTER)

        // constrain the preferred size so that the layout doesn't
        // change once the control panel is shown
        this.preferredSize = textArea.preferredSize
    }

    fun setColorSchemeList(file: File?) {
        if (!this.checkModifiedStateAndSaveIfNecessary()) {
            return
        }
        try {
            this.currentFile = file
            if (file != null) {
                this.schemes = SubstanceSkin.getColorSchemes(file.toURI().toURL())
            } else {
                this.schemes = SubstanceSkin.ColorSchemes()
            }
            // switch to the list view
            val cl = cardPanel.layout as CardLayout
            cl.show(cardPanel, LIST)
            // let the view know that the model has changed
            schemeListModel.fireContentsChanged()
            // update the buttons state
            if (file == null) {
                moveUpButton.isEnabled = false
                moveDownButton.isEnabled = false
                deleteButton.isEnabled = false
                renameButton.isEnabled = false
            }
            this.isModified = false
        } catch (exc: Exception) {
            exc.printStackTrace()
        }

    }

    fun checkModifiedStateAndSaveIfNecessary(): Boolean {
        if (!this.isModified) {
            return true
        }

        // does the user want to save modifications?
        val fileName = if (this.currentFile == null) "Unsaved" else this.currentFile!!.name

        val userSelection = this.showConfirmDialogInWindow(
                "Do you want to save the changes to '$fileName'?", "Modified contents",
                OptionPaneConfirmType.YES_NO_CANCEL_OPTION_TYPE)
        if (userSelection == JOptionPane.CANCEL_OPTION) {
            return false
        }
        if (userSelection == JOptionPane.YES_OPTION) {
            if (this.currentFile != null) {
                // ask to save
                this.save()
            } else {
                this.saveAs()
            }
        }
        return true
    }

    private fun getNewColorSchemeName(nameToStartWith: String?): String? {
        var result: String? = this.showInputDialogInWindow(
                "Type the color scheme name", nameToStartWith) ?: return null

        if (result == nameToStartWith) {
            return result
        }

        result = result!!.trim { it <= ' ' }

        // check characters
        val pattern = Pattern.compile("[a-zA-Z ]+")
        val matcher = pattern.matcher(result)
        if (!matcher.matches()) {
            this.showMessageDialogInWindow("Only use letters and spaces", "Name invalid",
                    OptionPaneMessageType.ERROR_MESSAGE_TYPE)
            return null
        }

        // check for uniqueness
        val existing = this.schemes!!.get(result)
        if (existing != null) {
            this.showMessageDialogInWindow("Name already exists", "Name clash",
                    OptionPaneMessageType.ERROR_MESSAGE_TYPE)
            return null
        }
        return result
    }

    fun save() {
        this.saveAs(this.currentFile)
    }

    fun saveAs() {
        // ask for a filename for the save
        val fileName = this.fileNameForSaving
        if (fileName != null) {
            val file = File(fileName)
            this.saveAs(file)
        }
    }

    private fun saveAs(file: File?) {
        val printStream = PrintStream(FileOutputStream(file!!))
        printStream.use {
            for (i in 0 until this.schemes!!.size()) {
                val colorScheme = this.schemes!!.get(i)
                val encodedColorScheme = colorScheme.toString()
                printStream.println(encodedColorScheme)
                printStream.println()
            }
        }
        this.currentFile = file
        this.isModified = false
    }

    fun updateColorScheme(colorScheme: SubstanceColorScheme) {
        this.schemes!!.replace(colorScheme.displayName, colorScheme)
    }

    internal inner class ColorSchemeListModel : AbstractListModel<SubstanceColorScheme>() {
        override fun getSize(): Int {
            return if (schemes == null) 0 else schemes!!.size()
        }

        override fun getElementAt(index: Int): SubstanceColorScheme? {
            return schemes?.get(index)
        }

        fun fireContentsChanged() {
            this.fireContentsChanged(schemeList, 0, size)
        }
    }

    internal inner class ColorSchemeListRenderer : SubstanceDefaultListCellRenderer() {
        override fun getListCellRendererComponent(list: JList<*>, value: Any?, index: Int,
                isSelected: Boolean, cellHasFocus: Boolean): Component {
            val scheme = value as SubstanceColorScheme?
            return super.getListCellRendererComponent(list, scheme!!.displayName, index,
                    isSelected, cellHasFocus)
        }
    }

    companion object {
        internal val LIST = "List"
        internal val INSTRUCTIONAL = "Instructional"
    }
}
